ludwig/models/retrieval.py

```diff
@@ -0,0 +1,225 @@
+import hashlib
+import json
+import logging
+import os
+from typing import Any, Dict, List, Optional, TYPE_CHECKING, Union
+
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+try:
+    import faiss
+except ImportError:
+    faiss = None
+
+if TYPE_CHECKING:
+    from ludwig.backend.base import Backend
+
+from ludwig.utils.batch_size_tuner import BatchSizeEvaluator
+from ludwig.utils.torch_utils import get_torch_device
+
+
+def df_checksum(df: pd.DataFrame) -> str:
+    return hashlib.sha1(pd.util.hash_pandas_object(df).values).hexdigest()
+
+
+class RetrievalModel:
+    def create_dataset_index(self, df: pd.DataFrame, backend: "Backend", columns_to_index: Optional[List[str]] = None):
+        """Creates an index for the dataset.
+
+        If `columns_to_index` is None, all columns are indexed. Otherwise, only the columns in `columns_to_index` are
+        used for indexing, but all columns in `df` are returned in the search results.
+        """
+        raise NotImplementedError
+
+    def search(
+        self, df, backend: "Backend", k: int = 10, return_data: bool = False
+    ) -> Union[List[int], List[Dict[str, Any]]]:
+        """Retrieve the top k results for the given query.
+
+        If `return_data` is True, returns the data associated with the indices. Otherwise, returns the indices.
+        """
+        raise NotImplementedError
+
+    def save_index(self, name: str, cache_directory: str):
+        """Saves the index to the cache directory."""
+        raise NotImplementedError
+
+    def load_index(self, name: str, cache_directory: str):
+        """Loads the index from the cache directory."""
+        raise NotImplementedError
+
+
+class RandomRetrieval(RetrievalModel):
+    """Random retrieval model.
+
+    Gets k random indices from the dataset regardless of the query.
+    """
+
+    def __init__(self, **kwargs):
+        self.index = None
+        self.index_data = None
```

Comment:
Should we also maintain a checksum here as well since we are writing the CSV form of the df + indices to disk? That way if the underlying data changes, we can recompute a new random index and also just return a pre-existing index if we've already saved it. I know the computation cost of Random retrieval is practically negligible, but it might be useful for consistency